flux-middle/src/rty/mod.rs

```diff
@@ -262,7 +262,7 @@ pub enum BaseTy {
     Tuple(List<Ty>),
     Array(Ty, Const),
     Never,
-    Closure(DefId),
+    Closure(DefId, List<Ty>),
```

Comment:
Do you mean our `ClosureSubsts` should be e.g. an alias/wrapper around (our) `Substs`?

Reply:
At any rate, what is the rationale for duplicating `ClosureSubsts`? The only thing we care about is the `U` -- the rest are random type params/lifetimes that don't seem relevant?

Reply:
Finally, note that we need to use the actual arguments (and *not* the `substs` lowered from `rustc`) when actually creating the `rty` for the aggregate:  

https://github.com/flux-rs/flux/blob/28518970fb7fb03e3249d7174149a1fb4ab5176e/flux-refineck/src/checker.rs#L696-L706

The old `_substs` seems to not be required? If we did want to cart it around, I think we'd have to:

* convert (with some trivial refinements?) the first `n-1` elements of `substs`
* plug in the tupled-up types from the actual operands as the last `n`-th element of `substs`
* create a `Closure(..., substs)`

[ anyways, I guess the main point is: not clear what the value of carrying those first n-1 arguments is... perhaps you have some rationale? ]

Reply:
The Substs also carries the generic parameter from the surrounding context, I'm not sure when we'll need them, but I think is good to keep them around. This is not the important part though. In the common case, when you are capturing references (i.e., the closure is not `move`), we should be able to weaken the type for mutable references. So the way I would do it is to generate default refinements with holes for the original `_substs` and then check subtyping. In case the closure is `move`, as an optimization, we can reuse the types for the `args` and avoid subtyping.

Reply:
Do you have a small test case in mind that we can use to test this?

Reply:
I don't know if this will work out of the box with the suggestion I'm making but I'm thinking something like

```rust
let mut x = 0;
let opt = Some(1).map(|_| { x += 1; 0 });
assert(x >= 0)
```

Reply:
On second thought, I may be conflating things. The issue with defining a `ClosureSubsts`  is independent. We may just need to modify `pack_closure_operands` to handle the case `TyKind::Ptr(PtrKind::Mut, ..)`